server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -215,7 +216,15 @@ static DfsKnnResults singleKnnSearch(Query knnQuery, int k, Profilers profilers,
                 topDocsCollectorManager,
                 CollectorResult.REASON_SEARCH_TOP_HITS
             );
-            topDocs = searcher.search(knnQuery, ipcm);
+            // When profiling, we want to make sure that all tasks are done on the task executor
+            // Otherwise we can't guarantee that the `knnProfiler` isn't being hit by more than one thread
+            if (searcher.getTaskExecutor() != null) {
+                List<Callable<TopDocs>> tasks = new ArrayList<>(1);
+                tasks.add(() -> searcher.search(knnQuery, ipcm));
```

Comment:
Collections.singletonList?

Reply:
The TaskExecutor has a protection mechanism to verify that a certain operation being run on the executor is not further parallelized using additional threads from the executor, as that could cause a deadlock. With that I am not sure whether calling search directly on the executor would work. Wouldn't the search offload the single slice execution to the executor, and wouldn't that fail?

Reply:
> Wouldn't the search offload the single slice execution to the executor, and wouldn't that fail?

I honestly don't know. The tests pass with this change. What can we test to ensure this is protected?

I am unfamiliar with all the task parallelism code. I just know the way it currently is (profiling & parallel) is broken :)

Reply:
The protection mechanism works fine and that is what prevents issues: https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java#L147 . If it detects that we are already within the executor , we won't fork further but rather call `Runnable#run` on the current thread.

Reply:
I opened #108041 with a fix to the profile code. I also understood why your stop-gap fixed worked: by using the search workers executor to execute the search, you implicitly disable parallelism, because task executor will detect that it's already within the executor itself and will refuse to parallelize further.

Reply:
thank you @javanna !!!